llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Reg.isVirtual())
+          continue;
+        if (VRM->hasRequiredPhys(Reg))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI->isVecOrAccRegClass(MRI->getRegClass(Reg)))
+          continue;
+        //  Subregisters can reuse Constants in the high part of the
+        //  register, therefore renaming a reused physical register can
+        //  prohibit the elimination of a move of constants, thus degrading
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs)) {
+          MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+
+          if (replaceReg(Reg, ForbiddenPhysRegs)) {
+            LLVM_DEBUG(dbgs() << MI);
+            Modified = true;
+            ReplacementRegs.push_back(
+                std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg).id()));
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM->getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
+      }
+    }
+    // dump Reassignment
+    LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+               dbgs() << "\n");
+    for (std::pair<MCRegister, MCRegister> &RegPair : ReplacementRegs) {
+      LLVM_DEBUG(dbgs() << printReg(RegPair.first, TRI, 0, MRI) << " -> "
+                        << printReg(RegPair.second, TRI, 0, MRI) << "\n");
+    }
+    LLVM_DEBUG(dbgs() << "\n");
+  }
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWAWCandidate(
+    const Register &Reg, const MachineInstr &MI,
+    const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+    const SmallVector<MCPhysReg, 16> &UsedVirtRegs) {
+  const bool FoundPhyReg =
+      UsedPhyRegs.end() != find(UsedPhyRegs, VRM->getPhys(Reg));
+  const bool FoundVirtReg =
+      UsedVirtRegs.end() == find(UsedVirtRegs, Register::virtReg2Index(Reg));
+
+  return MRI->getRegClass(Reg) && FoundPhyReg && FoundVirtReg && !MI.isCopy();
+}
+
+SmallSet<MCPhysReg, 16>
+AIEWawRegRewriter::getMBBForbiddenReplacements(const MachineBasicBlock *MBB) {
+  SmallSet<MCPhysReg, 16> BlockedPhysRegs;
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      addAliasRegs(BlockedPhysRegs, getPhysReg(Op));
+    }
+  }
+
+  LLVM_DEBUG(dbgs() << "Forbidden Physical Register of the bb."
+                    << MBB->getNumber() << "\n");
+  for (auto &Reg : BlockedPhysRegs) {
+    LLVM_DEBUG(dbgs() << printReg(Reg, TRI, 0, MRI) << " ");
+  }
```

Comment:
Nit: Maybe wrap that whole loop in a `LLVM_DEBUG` macro?

Reply:
moved the for loop into the LLVM_DEBUG pragma